### ISSUE TITLE
ParaformerLarge: BF16 8-bit optimizer throws 'Source array was not long enough' at 661M-param scale

### ISSUE THREAD

**Issue by ooples** [ooples/AiDotNet#1767]
### Summary
When `ParaformerLarge<T>` trains at its default (foundation) scale — a warm-up forward reports **~661M trainable parameters** (`GetParameters().Length = 661,219,029`) — the auto-selected BF16 8-bit Adam path throws:

```
System.ArgumentException : Source array was not long enough. Check the source index, length, and the array's lower bounds. (Parameter 'sourceArray')
   at AiDotNet.Tensors.Helpers.TensorAllocator.RentUninitialized[T](Int32[] shape)
   at AiDotNet.Tensors.Engines.CpuEngine.TensorMultiplyScalar[T](Tensor`1 tensor, T scalar)
   at AiDotNet.Optimizers.Adam8BitOptimizer`3.Step(TapeStepContext`1 context)   // BF16 moment path, ~line 605
   at AiDotNet.NeuralNetworks.NeuralNetworkBase`1.TrainWithTape(...)
   at AiDotNet.SpeechRecognition.AlibabaASR.ParaformerLarge`1.Train(...)
```

### Context
- Surfaced by the Generated Layers N-S ModelFamily shard (PR #1744). ParaformerLarge is now tagged `[Category=HeavyTimeout]` and deferred to the nightly heavy lane (its training invariants also time out at the 120-180s CPU gate budget), so it no longer blocks the PR gate — this issue tracks the underlying optimizer error for the nightly lane.
- The model auto-selects `Adam8BitOptimizer` with `UseBFloat16MomentStorage=true` (via `NeuralNetworkBase.ShouldUseBFloat16Optimizer`) because it crosses the large-model parameter threshold. The error is in that BF16 `Step` path.
- **Not reproducible locally** on a CPU dev box: the 100-iter training times out (~180s) before reaching the failing step, so this only manifests on the faster CI runner. A per-step length-invariant probe (`grad.Length==param.Length`, `mBf16.Length==param.Length`, `product(grad._shape)==grad.Length`) did NOT trip during the (partial) local run, so the mismatch is not a simple param/grad length disagreement in the common case — likely a pooled-buffer/scale interaction specific to 661M-element tensors.

### Likely area to investigate
`Adam8BitOptimizer.Step` BF16 branch (the `TensorMultiplyScalar(grad, 1-beta1)` → `AutoTensorCache.RentOrAllocate` → `TensorAllocator.RentUninitialized` chain) at ~661M-element scale, and/or `ShouldUseBFloat16Optimizer` engaging BF16 for a param count where the rent/pool path overflows. Cross-reference PR #1745 (BF16-Adam fused-compatibility), which last touched this path.

Raised while greening Generated Layers N-S in PR #1744.

**Comment by ooples** [comment]
Investigation update:

- **Disproved the int/byte-overflow hypothesis.** `TensorMultiplyScalar` on a 550M-element `float` tensor (2.2 GB — already past `int.MaxValue` bytes) with a fresh `TensorArena` runs fine. The generic large-tensor rent/allocate/construct path is correct above 2 GB, so this is **not** a simple size overflow.
- **It's arena-state-dependent.** The failure needs the arena/pool state accumulated over a full foundation-scale training step; a fresh-arena isolated repro doesn't trigger it. That matches "only on the fast CI runner" — the local 100-iter training times out (~180 s) before the failing step, so it can't be reproduced on a CPU dev box.

To capture the true cause, I opened a **diagnostic** PR in the Tensors repo (ooples/AiDotNet.Tensors#725): large (≥100M-element) `RentUninitialized` failures now rethrow with the requested shape/size and the live arena + pool state (ring sizes, per-bucket `(type,size)` keys **with backing-array lengths** — a length ≠ key-size is the suspected signature). Once that ships, the nightly HeavyTimeout `ParaformerLarge` run will surface the real failing location + arena state, enabling the actual fix.

This issue stays open until the real fix lands (the diagnostic doesn't fix it).

**Comment by ooples** [comment]
Triage note (PR #1789 sweep): reviewed against the #1789 branch. There is related activity on that branch but **no verified fix for what this issue actually asks for**, so it is deliberately left open and not linked for auto-close. Re-check after #1789 merges.

**Comment by ooples** [comment]
### Re-audit against master (`origin/master` @ `a22d7e78bf` (2026-09-05))

**Why this correction exists.** The `Triage note (PR #1789 sweep)` comment on this issue was produced by checking the **#1789 branch**, which was cut on **2026-07-05**. That sweep ran on **2026-08-14**, the day before #1789 merged. Any fix that reached `master` in that 5.5-week window *outside* #1789 was invisible to it, so those verdicts are false-negative-prone. Re-verified below against master itself.

**Verdict: OPEN - and the model has no test coverage**

Re-checked on master. `ParaformerLarge` is **not referenced anywhere in the test tree** - no model-family or integration test exercises it.

So the BF16 8-bit optimizer failure at 661M-param scale is neither fixed nor detectable: nothing in CI would catch it if it regressed further, and nothing would confirm a fix. The missing coverage is arguably the more important half of this issue.

Recommend adding a test before attempting the fix. Triaged v1 Tier 3 (post-v1).

---
*Posted by an automated re-audit of all 45 open issues against master on 2026-09-05. Every figure above was measured on the tree or read from CI run artifacts; none is carried over from a previous comment.*
